src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -385,175 +376,186 @@
-
-    (start-rate-monitor rate-monitor-ch run-interval commands-per-puppet-run)
-
-    (dotimes [_ threads]
-      (start-command-sender base-url command-send-ch rate-monitor-ch))
-
-    (when-not catalogs (log/info (trs "No catalogs specified; skipping catalog submission")))
-    (when-not reports (log/info (trs "No reports specified; skipping report submission")))
-    (when-not facts (log/info (trs "No facts specified; skipping fact submission")))
-
-    (if nummsgs
-      (do
-        (submit-n-messages hosts nummsgs rand-perc command-send-ch)
-        (dotimes [_ threads]
-          (>!! command-send-ch :stop))
-        (close! command-send-ch))
-
-      (let [close-to-stop-ch (chan)
-            mq (chan (amq-broker-buffer "benchmark-mq" "benchmark-endpoint"))]
-
-        ;; be sure to clear the temp queue on exit
-        (.addShutdownHook (Runtime/getRuntime) (Thread. #(close! mq)))
-
-        ;; initial queue population
-        (future
-          (println "Populating the queue in the background")
-          (dotimes [n numhosts]
-            (let [h (-> (make-host n)
-                        (assoc :lastrun (rand-lastrun run-interval)))]
-              (>!! mq h)))
-
-          (println "... finished filling the queue, continuing with the simulation."))
-
-        (let [run-interval-minutes (time/in-minutes run-interval)
-              hosts-per-second (/ numhosts (* run-interval-minutes 60))
-              rate-limited-mq-out (chan)]
-          (relay-messages-at-rate mq rate-limited-mq-out hosts-per-second)
-          (pipeline simulation-threads
-                    mq
-                    (map #(update-host % rand-perc (time/now) command-send-ch))
-                    rate-limited-mq-out)
-          (go
-            (<! close-to-stop-ch)
-            (mapv close! [rate-limited-mq-out mq rate-monitor-ch command-send-ch])))
-        close-to-stop-ch))))
-
-(defn chan? [x]
-  (satisfies? clojure.core.async.impl.protocols/Channel x))
-
+                                   (if facts 1 0))
+        temp-dir (get-path (or (System/getenv "TMPDIR")
+                               (System/getProperty "java.io.tmpdir")))
+
+        ;; channels
+        mq-ch (chan (message-buffer temp-dir numhosts))
+        _ (.addShutdownHook (Runtime/getRuntime) (Thread. #(async/close! mq-ch)))
+
+        command-send-ch (chan)
+        rate-monitor-ch (chan)
+
+        ;; processes
+        rate-monitor-finished-ch (start-rate-monitor rate-monitor-ch run-interval commands-per-puppet-run)
+        command-sender-finished-ch (start-command-sender base-url
+                                                         (if nummsgs
+                                                           (async/take (* numhosts nummsgs) command-send-ch)
+                                                           command-send-ch)
+                                                         rate-monitor-ch
+                                                         threads)
```

Comment:
I think simulation-threads was added later. You want me to make that an arg too?